### PR TITLE
coprocessor/dag/expr: add OctInt

### PR DESCRIPTION
**Pull request by yjhmelody**
Signed-off-by: yjh <yjh465402634@gmail.com>

## What have you changed? (mandatory)
implement builtin function: OctInt in coprocessor.
#3275

## How has this PR been tested? (mandatory)
`make expression`

### CODE REVIEW

**Comment by siddontang** [comment]
Hi @yjhmelody 

Please sign off your all commits. See https://github.com/tikv/tikv/blob/master/CONTRIBUTING.md#signing-off-the-commit

If you have already pushed some commits before, you can use `git rebase -i master --signoff` to resign them off.

File: src/coprocessor/dag/expr/builtin_string.rs [hunk 1]
```diff
@@ -297,6 +307,43 @@ mod test {
...
+                Datum::Bytes(b"777777777777777777777".to_vec()),
+            ),
+            (
+                Datum::I64(i64::min_value()),
+                Datum::Bytes(b"1000000000000000000000".to_vec()),
+            ),
+            (
+                Datum::I64(-1),
+                Datum::Bytes(b"1777777777777777777777".to_vec()),
+            ),
```
Comment (spongedu) [inline src/coprocessor/dag/expr/builtin_string.rs]: What about add some corner cases?  For example:
```sql
mysql> select oct(18446744073709551616);
+---------------------------+
| oct(18446744073709551616) |
+---------------------------+
| 1777777777777777777777    |
+---------------------------+
1 row in set (0.00 sec)

mysql> select oct(-9223372036854775809);
+---------------------------+
| oct(-9223372036854775809) |
+---------------------------+
| 777777777777777777777     |
+---------------------------+
1 row in set (0.00 sec)

mysql> select oct(111111111111111111111111111111111111111111111111111111111111111);
+----------------------------------------------------------------------+
| oct(111111111111111111111111111111111111111111111111111111111111111) |
+----------------------------------------------------------------------+
| 1777777777777777777777                                               |
+----------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> select oct(18446744073709551614);
+---------------------------+
| oct(18446744073709551614) |
+---------------------------+
| 1777777777777777777776    |
+---------------------------+
1 row in set (0.00 sec)

mysql> select oct(18446744073709551615);
+---------------------------+
| oct(18446744073709551615) |
+---------------------------+
| 1777777777777777777777    |
+---------------------------+
1 row in set (0.00 sec)

```

Comment (yjhmelody) [reply src/coprocessor/dag/expr/builtin_string.rs]: I feel strange about  `111111111111111111111111111111111111111111111111111111111111111`, 
`#[allow(overflowing_literals)]` also cannot make it compile.
And `18446744073709551615` is Contradicted with one of my test data.

Comment (yjhmelody) [reply src/coprocessor/dag/expr/builtin_string.rs]:
```sql
+---------------------------+
| oct(18446744073709551616) |
+---------------------------+
| 1777777777777777777777    |
+---------------------------+
```
```sql
+---------------------------+
| oct(18446744073709551615) |
+---------------------------+
| 1777777777777777777777    |
+---------------------------+
```
Are they not contradictory?

Comment (spongedu) [reply src/coprocessor/dag/expr/builtin_string.rs]: @yjhmelody  yep, that's why we should test the corner cases, the behavior of ` format! ` is not always as we expected..

**Comment by DorianZheng** [comment]
@breeswish PTAL

File: src/coprocessor/dag/expr/builtin_string.rs (hunk 1 above)
Comment (yjhmelody) [reply src/coprocessor/dag/expr/builtin_string.rs]: I means why oct(18446744073709551616) == oct(18446744073709551615)
And ...
How it can do that

OCT(N)
- Returns a string representation of the octal value of N, where N is a longlong (BIGINT) number. This is equivalent to CONV(N,10,8). Returns NULL if N is NULL.

Comment (spongedu) [reply src/coprocessor/dag/expr/builtin_string.rs]: @yjhmelody Let's take a look into MySQL source code.  `oct` is implemented in [`sql/item_create.cc`](https://github.com/mysql/mysql-server/blob/913071c0b16cc03e703308250d795bc381627e37/sql/item_create.cc#L6784), which direct us to [`sql/item_strfunc.cc`](https://github.com/mysql/mysql-server/blob/913071c0b16cc03e703308250d795bc381627e37/sql/item_strfunc.cc#L3929 ).
Line 3966 is where the input string is converted to int:
```c++
      dec= (longlong) my_strntoull(res->charset(), res->ptr(), res->length(),
                                   from_base, &endptr, &err);
```
Following the function we can trace down to [`strings/ctype-simple.c`](https://github.com/mysql/mysql-server/blob/b93c1661d689c8b7decc7563ba15f6ed140a4eb6/strings/ctype-simple.cc#L1221):
```c++
  for (; shift > 0; shift--, ull *= 10) /* Left shift */
  {
    if (ull > CUTOFF) goto ret_too_big; /* Overflow, number too big */
  }
```

```c++
ret_too_big:
  *endptr = (char *)str;
  *error = MY_ERRNO_ERANGE;
  return unsigned_flag ? ULLONG_MAX
                       : negative ? (ulonglong)LLONG_MIN : (ulonglong)LLONG_MAX;
}
```
These parts indicate that if this function return INT_MAX/INT_MIN if the input is overflow. 
so oct(18446744073709551616) == oct(18446744073709551615) = oct(1111111111111111111111111111111111111111111111).

File: src/coprocessor/dag/expr/builtin_string.rs [hunk 2]
```diff
@@ -45,6 +45,61 @@ impl ScalarFunc {
...
+    pub fn oct_int<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &[Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let input = try_opt!(self.children[0].eval_string(ctx, row));
+        let mut num: u64 = 0;
+        let mut i: usize = 1;
+        let mut overflow = false;
+        let neg = match input[0] {
```
Comment (spongedu) [inline src/coprocessor/dag/expr/builtin_string.rs]: What about use `self.children[0].eval_int(ctx, row)` to get input parameter and deal with overflow, like [implementation of builtin `bit_count`](https://github.com/tikv/tikv/blob/397b4a6e9ea6b5dd244247f726d031a4b53d059c/src/coprocessor/dag/expr/builtin_other.rs#L23) ?

Comment (yjhmelody) [reply src/coprocessor/dag/expr/builtin_string.rs]: The `is_overflow`, Can diff the max_value overflow and min_value overflow?

Comment (spongedu) [reply src/coprocessor/dag/expr/builtin_string.rs]: @yjhmelody  I take some extra experiments, and found that it seems that MySQL always treats input as unsigned. If so, there's no need to distinguish max/min value when overflow. 
The following are my experiments:
```sql
mysql> select oct(9223372036854775807);
+--------------------------+
| oct(9223372036854775807) |
+--------------------------+
| 777777777777777777777    |
+--------------------------+
1 row in set (0.00 sec)

mysql> select oct(-9223372036854775809);
+---------------------------+
| oct(-9223372036854775809) |
+---------------------------+
| 777777777777777777777     |
+---------------------------+
1 row in set (0.00 sec)

mysql> select oct(9223372036854775806);
+--------------------------+
| oct(9223372036854775806) |
+--------------------------+
| 777777777777777777776    |
+--------------------------+
1 row in set (0.00 sec)

mysql> select oct(-9223372036854775810);
+---------------------------+
| oct(-9223372036854775810) |
+---------------------------+
| 777777777777777777776     |
+---------------------------+
1 row in set (0.00 sec)

mysql> select oct(-9223372036854775808);
+---------------------------+
| oct(-9223372036854775808) |
+---------------------------+
| 1000000000000000000000    |
+---------------------------+
1 row in set (0.00 sec)

mysql> select oct(9223372036854775808);
+--------------------------+
| oct(9223372036854775808) |
+--------------------------+
| 1000000000000000000000   |
+--------------------------+
1 row in set (0.01 sec)

mysql> select oct(-11111111111111111111111111111111);
+----------------------------------------+
| oct(-11111111111111111111111111111111) |
+----------------------------------------+
| 1777777777777777777777                 |
+----------------------------------------+
1 row in set (0.00 sec)

mysql> select oct(11111111111111111111111111111111);
+---------------------------------------+
| oct(11111111111111111111111111111111) |
+---------------------------------------+
| 1777777777777777777777                |
+---------------------------------------+
1 row in set (0.00 sec)

```

I'll check the MySQL source code to confirm this. You can take a look at this as well. By the way, shall we add these cases into tests?

Comment (yjhmelody) [reply src/coprocessor/dag/expr/builtin_string.rs]: I have a question. Should I rewrite the tests just as `builtin bit_count ` ?  I found it has different Datum type to test.  While I only test the Datum::Bytes because of the `oct(11111111111111111111111111111111)`
Let me check your meaning, if the number overflow, we consider it 64 binary numbers all 1.  And, when we parse it to oct, we only map it binary value to a determined oct value.
If so, it seem that these are only little work need to do.

Comment (spongedu) [reply src/coprocessor/dag/expr/builtin_string.rs]: 1. Maybe write the tests just as `builtin bit_count` is better because this is builtin `OctInt`. Current form of tests is more suitable for `OctString`, I think. 
2. We'd better check the MySQL source code first. I'm not sure now, as well

Comment (spongedu) [reply src/coprocessor/dag/expr/builtin_string.rs]: @yjhmelody Hi, I've checked MySQL source code. in [sql/item_strfunc.cc](https://github.com/mysql/mysql-server/blob/913071c0b16cc03e703308250d795bc381627e37/sql/item_strfunc.cc#L3970), after getting an `ulonglong` input, the following line is executed to transfer `ulonglong` into `string`:
```c++
     ptr= longlong2str(dec, ans, to_base))
```

`longlong2str` is defined in [include/m_string.h](https://github.com/mysql/mysql-server/blob/913071c0b16cc03e703308250d795bc381627e37/include/m_string.h#L222) and implemented in [strings/longlong2str.c](https://github.com/mysql/mysql-server/blob/913071c0b16cc03e703308250d795bc381627e37/strings/longlong2str.c#L50).
you can see in [line 63](https://github.com/mysql/mysql-server/blob/913071c0b16cc03e703308250d795bc381627e37/strings/longlong2str.c#L63) that there's a conversion for negative inputs:
```c
      /* Avoid integer overflow in (-val) for LLONG_MIN (BUG#31799). */
      uval = (ulonglong)0 - uval;
```

Comment (yjhmelody) [reply src/coprocessor/dag/expr/builtin_string.rs]: So, what's the conclusion?
I was originally based FormatUint which used in tidb and is similar to the definition which you showed me.
I'm a little confused about the version of C.

Comment (spongedu) [reply src/coprocessor/dag/expr/builtin_string.rs]: TiDB make `OctInt` simple because big integer-literal are treated as string constant and is processed in `octString`.   I think we can refine the current implementation a little, convert it into `OctString`, so we can make `OctInt` more simple as your original version.

**Comment by huachaohuang** [comment]
@DorianZheng PTAL

**Comment by huachaohuang** [comment]
@crazycs520 PTAL

**Comment by DorianZheng** [comment]
@yjhmelody Maybe you can implement `OctString` and implement `OctInt` based on it.

**Comment by yjhmelody** [comment]
@DorianZheng  Hi! Such like this?

File: src/coprocessor/dag/expr/builtin_string.rs [hunk 3]
```diff
@@ -45,6 +45,74 @@ impl ScalarFunc {
...
     }
 
+    #[inline]
+    pub fn oct_int<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &[Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        match self.children[0].eval_int(ctx, row) {
+            Err(_e) => self.oct_string(ctx, row),
```
Comment (DorianZheng) [inline src/coprocessor/dag/expr/builtin_string.rs]: Please stay the same with TiDB implementation.
FYI: https://github.com/pingcap/tidb/blob/master/expression/builtin_string.go#L2507

If it's error, return it.

File: src/coprocessor/dag/expr/builtin_string.rs [hunk 4]
```diff
@@ -45,6 +45,74 @@ impl ScalarFunc {
...
+        let neg = match input[0] {
+            b'-' => true,
+            b'+' => false,
+            _ => {
+                i -= 1;
+                false
+            }
+        };
+
+        while i < input.len() {
```
Comment (DorianZheng) [inline src/coprocessor/dag/expr/builtin_string.rs]: Maybe you can use https://doc.rust-lang.org/std/primitive.str.html#method.parse to parse the string to u64

and deal with negative or overflow like this: https://github.com/pingcap/tidb/blob/master/expression/builtin_string.go#L2533

**Comment by DorianZheng** [comment]
@yjhmelody Thank you for the refine. I leave some comment. PTAL

**Comment by huachaohuang** [comment]
@yjhmelody feel free to reopen this if you are still interested, thanks.

**Comment by ice1000** [comment]
@huachaohuang OP cannot re-open PRs.

**Comment by huachaohuang** [comment]
Oops, sorry, maybe just tell me to reopen it.

**Comment by yjhmelody** [comment]
@huachaohuang  I'm sorry for long time not to update. Well, how to reopen it?

**Comment by huachaohuang** [comment]
Let me reopen it for you :)

**Comment by yjhmelody** [comment]
@DorianZheng hi! Long time not to update it. now I am a little confused about the code which I need to alter.

File: src/coprocessor/dag/expr/builtin_string.rs (hunk 4 above)
Comment (yjhmelody) [reply src/coprocessor/dag/expr/builtin_string.rs]: Hi! but how convert the Vec  to String and then can I use the parse.

File: src/coprocessor/dag/expr/builtin_string.rs (hunk 3 above)
Comment (AndreMouche) [reply src/coprocessor/dag/expr/builtin_string.rs]: +1

**Comment by AndreMouche** [comment]
This PR will be closed as it's a long time since the last update.
